### ISSUE TITLE
BUG: Disk checkpointing with `tape.timestepper()` causes incorrect BC evaluation at taping.

### ISSUE THREAD

**Issue by sghelichkhani** [firedrakeproject/firedrake#4206]
When using disk checkpointing with `tape.timestepper()` (i.e., with `SingleDiskStorageSchedule`), we observe inconsistent behavior in the forward solve even before invoking any functional or derivative evaluation. This appears specifically when updating a `DirichletBC` field during a time-stepping loop.

Consider the following reproducer:

```
from firedrake import *
from firedrake.adjoint import *
from checkpoint_schedules import SingleDiskStorageSchedule as scheduler
# from checkpoint_schedules import SingleMemoryStorageSchedule as scheduler

continue_annotation()
enable_disk_checkpointing()

tape = get_working_tape()
tape.enable_checkpointing(scheduler())


mesh = checkpointable_mesh(UnitSquareMesh(5, 5))
V = FunctionSpace(mesh, "CG", 2)
T = Function(V).interpolate(0.0)
u = TrialFunction(V)
v = TestFunction(V)
a = inner(grad(u), grad(v)) * dx
x = SpatialCoordinate(mesh)
F = Function(V)
control = Control(F)
F.interpolate(sin(x[0] * pi) * sin(2 * x[1] * pi))
L = F * v * dx

bcs = [DirichletBC(V, T, (1,))]
uu = Function(V)

for i in tape.timestepper(iter(range(3))):
    T.assign(T + 1.0)
    solve(a == L, uu, bcs=bcs, solver_parameters={"ksp_type": "preonly", "pc_type": "lu"})

obj = assemble(uu * uu * dx)
pause_annotation()
print(f"Objective at the end of taping:{obj}")

rf = ReducedFunctional(obj, control)
print(f"Objective by calling functional: {rf(F)}")

```

would produce inconsistent values:

```
Objective at the end of taping:4.000901957850139
Objective by calling functional: 1.0009019578501521
```

whereas NOT checkpointing to disk, by using `SingleMemoryStorageSchedule` as scheduler (line 4) gives consistent AND different values:

```
Objective at the end of taping:9.000901957850123
Objective by calling functional: 9.000901957850061
```

**Comment by jrmaddison** [comment]
This is my understanding based on some reading of the pyadjoint and Firedrake code, but perhaps others more familiar with the pyadjoint internals might correct:

`BlockVariable.output` is a symbolic variable and, during the _initial_ forward run (the one used to build the tape), also references a value. This is natural in UFL where symbolic variables (`Coefficient`s) typically also carry values (`Function`s and `Cofunction`s). However when actually using the tape `BlockVariable.output` is used _only_ as a symbolic variable, and the value is accessed via `BlockVariable.saved_output`. e.g. you can see this in `GenericSolveBlock._create_F_form`, which builds a new form using `saved_output`s. The symbolic variable never changes, but the value in `BlockVariable.saved_output` can be freed from RAM and then rematerialized using a checkpointing schedule.

As an aside: this construction is what leads to issue https://github.com/dolfin-adjoint/pyadjoint/issues/169. While `BlockVariable.output` is only used as a symbolic variable by the adjoint, it still references a value, even though in principle the adjoint should no longer need it.

I think this means that an adjoint calculation should not modify any value referenced by `BlockVariable.output`, and so replacing

https://github.com/firedrakeproject/firedrake/blob/6a5f4b1d4b851f3518abdcdc78a57d0aad49981b/firedrake/adjoint_utils/dirichletbc.py#L46-L47

with

```
        if checkpoint is not None:
            bc = type(self)(self.function_space(), checkpoint.saved_output, self.sub_domain)
            # Hack so that bc._ad_create_checkpoint works
            bc.block = self.block
            return bc
```

should workaround the bug -- which it seems to in the reproducer. I'm not sure of the best way to avoid the hack in the above as I'm not very familiar with `FloatingType`s.

**Comment by dham** [comment]
I haven't had time to look at this in detail, but my understanding is that `BlockVariable.output` should only be used during taping (I think it's very badly named). `BlockVariable` should be storing everything that is needed during tape evaluation in attributes other than `output`.
